File: Data Structure & Algorithm/Summary List/Problem Collection/Medium/101. Symmetric Tree/101. Symmetric Tree.ts
// 101. Symmetric Tree

// Given the root of a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).

// Example 1:
// Input: root = [1,2,2,3,4,4,3]
// Output: true

// Example 2:
// Input: root = [1,2,2,null,3,null,3]
// Output: false

// Definition for a binary tree node.
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

// Recursion
// Time Complexity: O(n)
// Space Complexity: O(n)
export function isSymmetric(root: TreeNode | null): boolean {
  if (!root) {
    return true;
  }

  const recursion = (left: TreeNode | null, right: TreeNode | null): boolean => {
    if (!left && !right) {
      return true;
    }

    if (left && right) {
      if (left.val !== right.val) {
        return false;
      }
      return recursion(left.left, right.right) && recursion(left.right, right.left);
    } else {
      return false;
    }
  };

  return recursion(root.left, root.right);
}

File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Summary List/Code Random Thoughts/2nd round/Tree/101. Symmetric Tree/101. Symmetric Tree.go
// 101. Symmetric Tree

// Given the root of a binary tree, check whether it is a mirror of itself (i.e., symmetric around its center).

// Example 1:
// Input: root = [1,2,2,3,4,4,3]
// Output: true

// Example 2:
// Input: root = [1,2,2,null,3,null,3]
// Output: false

package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}

	var recursion func(left *TreeNode, right *TreeNode) bool
	recursion = func(left *TreeNode, right *TreeNode) bool {
		if left == nil && right == nil {
			return true
		} else if left != nil && right == nil {
			return false
		} else if left == nil && right != nil {
			return false
		} else {
			if left.Val != right.Val {
				return false
			}
			return recursion(left.Left, right.Right) && recursion(left.Right, right.Left)
		}
	}

	return recursion(root.Left, root.Right)
}
